src/std_extensions/quantum.rs

```diff
@@ -1,19 +1,168 @@
 //! Basic HUGR quantum operations
 
+use std::num::NonZeroU64;
+
 use smol_str::SmolStr;
 
-use crate::extension::prelude::{BOOL_T, QB_T};
+use crate::extension::prelude::{BOOL_T, ERROR_TYPE, QB_T};
 use crate::extension::{ExtensionId, SignatureError};
 use crate::std_extensions::arithmetic::float_types::FLOAT64_TYPE;
 use crate::type_row;
-use crate::types::type_param::TypeArg;
-use crate::types::FunctionType;
+use crate::types::type_param::{TypeArg, TypeArgError, TypeParam};
+use crate::types::{ConstTypeError, CustomCheckFailure, CustomType, FunctionType, Type, TypeBound};
+use crate::utils::collect_array;
+use crate::values::CustomConst;
 use crate::Extension;
 
 use lazy_static::lazy_static;
 
 /// The extension identifier.
 pub const EXTENSION_ID: ExtensionId = ExtensionId::new_unchecked("quantum");
+
+/// Identifier for the angle type.
+const ANGLE_TYPE_ID: SmolStr = SmolStr::new_inline("angle");
+
+fn angle_custom_type(precision_arg: TypeArg) -> CustomType {
+    CustomType::new(ANGLE_TYPE_ID, [precision_arg], EXTENSION_ID, TypeBound::Eq)
+}
+
+/// Angle type of a given precision (specified by the TypeArg).
+///
+/// This type is capable of representing angles that are multiples of 2π / 2^N where N is the
+/// precision.
+pub(super) fn angle_type(precision_arg: TypeArg) -> Type {
+    Type::new_extension(angle_custom_type(precision_arg))
+}
+
+/// The smallest forbidden precision.
+pub const PRECISION_BOUND: u8 = 54;
+
+const fn is_valid_precision(n: u8) -> bool {
+    n < PRECISION_BOUND
+}
+
+/// Type parameter for the precision of an angle.
+// SAFETY: unsafe block should be ok as the value is definitely not zero.
+#[allow(clippy::assertions_on_constants)]
+pub const PRECISION_TYPE_PARAM: TypeParam = TypeParam::bounded_nat(unsafe {
+    assert!(PRECISION_BOUND > 0);
+    NonZeroU64::new_unchecked(PRECISION_BOUND as u64)
+});
+
+/// Get the precision of the specified type argument or error if the argument is invalid.
+pub(super) fn get_precision(arg: &TypeArg) -> Result<u8, TypeArgError> {
+    match arg {
+        TypeArg::BoundedNat { n } if is_valid_precision(*n as u8) => Ok(*n as u8),
+        _ => Err(TypeArgError::TypeMismatch {
+            arg: arg.clone(),
+            param: PRECISION_TYPE_PARAM,
+        }),
+    }
+}
+
+pub(super) const fn type_arg(precision: u8) -> TypeArg {
+    TypeArg::BoundedNat {
+        n: precision as u64,
+    }
+}
+
+/// An angle
+#[derive(Clone, Debug, Eq, PartialEq, serde::Serialize, serde::Deserialize)]
+pub struct ConstAngle {
+    precision: u8,
+    value: u64,
+}
+
+impl ConstAngle {
+    /// Create a new [`ConstAngle`]
+    pub fn new(precision: u8, value: u64) -> Result<Self, ConstTypeError> {
```

Comment:
would it be useful to add a `from_f64(value: f64, precision: u8)`, with rounding down?

Reply:
It could be. Will add.

Reply:
Made a `from_radians()` method that takes an `f64`.